execgen: introduce // execgen:template directive for bools

jordanlewis opened the pull request:
Based on #49728.

This PR adds the // execgen:template<arg1, arg2> syntax for
concrete boolean template variables.

For example:

```
// execgen:inline
// execgen:template<templateParam>
func a (runtimeParam int, templateParam bool) int {
    if templateParam {
        return runtimeParam + 1
    } else {
        return runtimeParam - 1
    }
}

func main() {
    x := a(0, true)
    y := a(0, false)
    fmt.Println(x, y)
}
```

This code will print `1, -1` when eventually executed, but the
templateParam boolean will be evaluated at code generation time and not
appear in the generated code at all. Two variants of `a()` will be
inlined into their callsites - one with templateParam true, and one with
templateParam false.

Release note: None

jordanlewis committed:
execgen: introduce template directive for bools

This commit adds the // execgen:template<arg1, arg2> syntax for
concrete boolean template variables.

For example:

```
// execgen:inline
// execgen:template<templateParam>
func a (runtimeParam int, templateParam bool) int {
    if templateParam {
        return runtimeParam + 1
    } else {
        return runtimeParam - 1
    }
}

func main() {
    x := a(0, true)
    y := a(0, false)
    fmt.Println(x, y)
}
```

This code will print `1, -1` when eventually executed, but the
templateParam boolean will be evaluated at code generation time and not
appear in the generated code at all. Two variants of `a()` will be
inlined into their callsites - one with templateParam true, and one with
templateParam false.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/49939)
<!-- Reviewable:end -->

yuzefovich reviewed (commented):
> Reviewed 5 of 5 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49939)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @rohany)

___
*[pkg/sql/colexec/execgen/inline.go, line 110 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M99a6fd5UV0AEn3y4HW:-M99a6fd5UV0AEn3y4HX:blqccy5) ([raw file](https://github.com/cockroachdb/cockroach/blob/571b8ab0e733d340f629e224b81de7791e570159/pkg/sql/colexec/execgen/inline.go#L110)):*
> ```Go
> 
> 			// Replace template conditionals.
> 			body = monomorphizeTemplate(body, funcInfo, templateArgs)
> ```

Should `monomorphizeTemplate` call `replaceTemplateVars` itself?
___
*[pkg/sql/colexec/execgen/inline.go, line 245 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M99agV85wp71iOzFN3O:-M99agV85wp71iOzFN3P:b-fzr33p) ([raw file](https://github.com/cockroachdb/cockroach/blob/571b8ab0e733d340f629e224b81de7791e570159/pkg/sql/colexec/execgen/inline.go#L245)):*
> ```Go
> 					mustInline = true
> 				}
> 				if matches := templateRe.FindStringSubmatch(dec); matches != nil {
> ```

Where is `templateRe` defined?


<!-- Sent from Reviewable.io -->

rohany commented:
I'm trying to build this to play around -- does it compile? Some definitions are missing
```
pkg/sql/colexec/execgen/inline.go:34:37: undefined: funcInfo
pkg/sql/colexec/execgen/inline.go:166:30: undefined: replaceTemplateVars
pkg/sql/colexec/execgen/inline.go:245:19: undefined: templateRe
```

rohany commented:
What's an example of a function that we might want to extend this templating for other types on?

jordanlewis commented:
Ugh I must have missed a file. It compiles and all tests pass. will fix in a bit.

jordanlewis committed:
execgen: properly support inlining := calls

Previously, assigning the results of an inlined func with the := syntax
produced invalid Go code. Now, it makes sure to define the result
variables in the parent scope before entering the new inlined func
scope.

Release note: None

yuzefovich commented:
Could you please rebase given that first three commits have been merged?

jordanlewis commented:
Done.

yuzefovich commented:
Thanks!

yuzefovich reviewed (commented):
I have some questions and nits, but overall this looks great!

> Reviewed 10 of 10 files at r4, 11 of 11 files at r5, 2 of 2 files at r6, 2 of 2 files at r7, 2 of 2 files at r8, 2 of 2 files at r9.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-:-M9L-bEW2FJXlJc8tkfK:b-a3hrii)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jordanlewis and @rohany)

___
*[pkg/sql/colexec/hashjoiner_tmpl.go, line 41 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9L-1EI3l4ZTKVpMJ__:-M9L-1EI3l4ZTKVpMJ_a:bev5c7k) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8bd74498f409d389da32c192289b7a3a250a33e/pkg/sql/colexec/hashjoiner_tmpl.go#L41)):*
> ```Go
> 		currentID := hj.ht.probeScratch.headID[i]
> 
> 		var earlyReturn bool
> ```

Is there a reason we can't just use `return` instead of `break`s below? I think it definitely deserves a comment if so.
___
*[pkg/sql/colexec/hashjoiner_tmpl.go, line 269 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9L-RUwEAtS4bZEqLFe:-M9L-RUwEAtS4bZEqLFf:b9td1k1) ([raw file](https://github.com/cockroachdb/cockroach/blob/15c11df73f76bc5117a5919fea499c399bc9355d/pkg/sql/colexec/hashjoiner_tmpl.go#L269)):*
> ```Go
> 		}
> 	}
> ```

nit: seems like other functions do have an empty line before `return nResults`.
___
*[pkg/sql/colexec/hashtable_tmpl.go, line 520 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9Kh7v07MOkfRv0vowR:-M9Kh7v07MOkfRv0vowS:b-yaloiv) ([raw file](https://github.com/cockroachdb/cockroach/blob/e919deb166564bece70a23ac8acb957e7c77aaf0/pkg/sql/colexec/hashtable_tmpl.go#L520)):*
> ```Go
> 
> // execgen:inline
> // execgen:template<useSel>
> ```

Hm, this does look nice in general. Maybe originally I said that I'm in favor of introducing some kind of convention to highlight which variables are runtime and which are template just because in my mind I have a clear distinction between the two since I've spent a lot of time in the old framework. Jordan, I see your point that there is an execgen directive right above that says which variables are of template kind. I think I'm actually neutral on this question now.
___
*[pkg/sql/colexec/execgen/inline.go, line 35 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9K_hI95SVxFHrz7q3U:-M9K_hI95SVxFHrz7q3V:b-619og6) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b2ea0fc196f22c9f4788eb2945ca3ada1ef43de/pkg/sql/colexec/execgen/inline.go#L35)):*
> ```Go
> 	// Expand functions with // execgen:template.
> 	funcInfos := createTemplateFuncVariants(f)
> 	replaceTemplateCallsites(f, funcInfos)
> ```

nit: maybe `s/Callsites/CallSites/g`?
___
*[pkg/sql/colexec/execgen/template.go, line 119 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9Kbe5PEd0CpzA-EMS-:-M9Kbe5PEd0CpzA-EMS0:b-3r816p) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b2ea0fc196f22c9f4788eb2945ca3ada1ef43de/pkg/sql/colexec/execgen/template.go#L119)):*
> ```Go
> 								newElse := monomorphizeTemplate(n.Else, info, args)
> 								switch e := newElse.(type) {
> 								case *dst.BlockStmt:
> ```

Could you expand on why in the "if" case above we can only have `BlockStmt` and in the "else" case here we can have something else?
___
*[pkg/sql/colexec/execgen/template.go, line 223 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9Kcl1K-_YDrN-059ph:-M9Kcl1K-_YDrN-059pi:b39lhe8) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b2ea0fc196f22c9f4788eb2945ca3ada1ef43de/pkg/sql/colexec/execgen/template.go#L223)):*
> ```Go
> 			// Delete template params from runtime definition.
> 			newParamList := make([]*dst.Field, 0, len(n.Type.Params.List)-len(info.templateParams))
> 			for i, field := range n.Type.Params.List {
> ```

This seems very similar to a code snippet above, can we factor out a function and reuse it in both places? If it's not easy due to different types, then never mind.
___
*[pkg/sql/colexec/execgen/template.go, line 259 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9KdCQd8za_zFkI84_M:-M9KdCQd8za_zFkI84_N:b1m58nd) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b2ea0fc196f22c9f4788eb2945ca3ada1ef43de/pkg/sql/colexec/execgen/template.go#L259)):*
> ```Go
> 				},
> 			})
> 			// Extract the remaining execgen directives.
> ```

nit: I'd put new line before the comment, not after.
___
*[pkg/sql/colexec/execgen/template.go, line 263 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9KdkmV3Y4b8MSzqO8u:-M9KdkmV3Y4b8MSzqO8v:bhc166o) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b2ea0fc196f22c9f4788eb2945ca3ada1ef43de/pkg/sql/colexec/execgen/template.go#L263)):*
> ```Go
> 			directives := make([]string, 0)
> 			for _, s := range funcDecs.Start {
> 				if strings.HasPrefix(s, "// execgen") {
> ```

In `templateRe`, we have `\/\/ *` meaning that multiple spaces can match, but here we don't have that. Should we add a preprocessing step that would make all execgen directives use exactly one space after `//`?
___
*[pkg/sql/colexec/execgen/template.go, line 269 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9KdSZ7CYA6O7LkLcpF:-M9KdSZ7CYA6O7LkLcpG:b-b46px5) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b2ea0fc196f22c9f4788eb2945ca3ada1ef43de/pkg/sql/colexec/execgen/template.go#L269)):*
> ```Go
> 
> 			for i := range argsPossibilities {
> 				args := argsPossibilities[i]
> ```

super nit: could define `args` variable in `for` loop itself.
___
*[pkg/sql/colexec/execgen/template_test.go, line 27 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9KeUDbA9IQqr4UYZBK:-M9KeUDc2wYkreDSWIxX:brnhaxw) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b2ea0fc196f22c9f4788eb2945ca3ada1ef43de/pkg/sql/colexec/execgen/template_test.go#L27)):*
> ```Go
> 	}
> 
> 	res := generateAllTemplateArgs(arg)
> ```

Maybe also add that all 8 args are unique?


<!-- Sent from Reviewable.io -->

jordanlewis committed:
execgen: support early returns in inlined funcs

Early returns in inlined funcs are implemented using gotos.

Release note: None

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49939)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rohany and @yuzefovich)

___
*[pkg/sql/colexec/hashjoiner_tmpl.go, line 41 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9L-1EI3l4ZTKVpMJ__:-M9M0pGG8fKGjExlZqXI:b-qpnf50) ([raw file](https://github.com/cockroachdb/cockroach/blob/b8bd74498f409d389da32c192289b7a3a250a33e/pkg/sql/colexec/hashjoiner_tmpl.go#L41)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Is there a reason we can't just use `return` instead of `break`s below? I think it definitely deserves a comment if so.
</blockquote></details>

Without something special, you can't early-return from an inlined function.

But then I realized Go has gotos. I added another commit that implements early returns in inlined functions with gotos, which removes the requirement for this change, thanks for the suggestion.
___
*[pkg/sql/colexec/hashtable_tmpl.go, line 520 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9Kh7v07MOkfRv0vowR:-M9M16ctBNPiGeEgDp0Y:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/e919deb166564bece70a23ac8acb957e7c77aaf0/pkg/sql/colexec/hashtable_tmpl.go#L520)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Hm, this does look nice in general. Maybe originally I said that I'm in favor of introducing some kind of convention to highlight which variables are runtime and which are template just because in my mind I have a clear distinction between the two since I've spent a lot of time in the old framework. Jordan, I see your point that there is an execgen directive right above that says which variables are of template kind. I think I'm actually neutral on this question now.
</blockquote></details>

:thumbsup: 
___
*[pkg/sql/colexec/execgen/template.go, line 119 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9Kbe5PEd0CpzA-EMS-:-M9M2T0y8guoO1vd924T:b-5snmhf) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b2ea0fc196f22c9f4788eb2945ca3ada1ef43de/pkg/sql/colexec/execgen/template.go#L119)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Could you expand on why in the "if" case above we can only have `BlockStmt` and in the "else" case here we can have something else?
</blockquote></details>

That's how the AST is defined in Go. An If statement always contains a Body (a block stmt) and optionally an Else (a generic stmt). An if /else if / else is actually 2 nested if statements, where the else of the first if contains the second if/else, if that makes sense.

https://astexplorer.net/#/gist/668b53ce1504c5f02ebf56fe778b6717/26076b4bfe01ea86ad0b1b025cdd9a7b0be6fee4
___
*[pkg/sql/colexec/execgen/template.go, line 223 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9Kcl1K-_YDrN-059ph:-M9M3NKX2u_MAen3ub_V:b484l1z) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b2ea0fc196f22c9f4788eb2945ca3ada1ef43de/pkg/sql/colexec/execgen/template.go#L223)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

This seems very similar to a code snippet above, can we factor out a function and reuse it in both places? If it's not easy due to different types, then never mind.
</blockquote></details>

Good idea, but yeah, the types are subtly different so it doesn't feel worth it (sigh)
___
*[pkg/sql/colexec/execgen/template.go, line 263 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9KdkmV3Y4b8MSzqO8u:-M9M6eqf1bNnbHgz0HvZ:b-3gt9oz) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b2ea0fc196f22c9f4788eb2945ca3ada1ef43de/pkg/sql/colexec/execgen/template.go#L263)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

In `templateRe`, we have `\/\/ *` meaning that multiple spaces can match, but here we don't have that. Should we add a preprocessing step that would make all execgen directives use exactly one space after `//`?
</blockquote></details>

Good point, I guess I would prefer to get rid of the thing that templateRe does and require a single space. I did that instead.
___
*[pkg/sql/colexec/execgen/template.go, line 269 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9KdSZ7CYA6O7LkLcpF:-M9M6d9r5OjwrOUNtvnv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b2ea0fc196f22c9f4788eb2945ca3ada1ef43de/pkg/sql/colexec/execgen/template.go#L269)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

super nit: could define `args` variable in `for` loop itself.
</blockquote></details>

Done.
___
*[pkg/sql/colexec/execgen/template_test.go, line 27 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9KeUDbA9IQqr4UYZBK:-M9M70DI0a-zaUcII3n6:bm0sxk) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b2ea0fc196f22c9f4788eb2945ca3ada1ef43de/pkg/sql/colexec/execgen/template_test.go#L27)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Maybe also add that all 8 args are unique?
</blockquote></details>

Thanks, I had meant to do this but forgot. This is a lame test :) Fixed


<!-- Sent from Reviewable.io -->

asubiotto reviewed (commented):
:partying_face: 

> Reviewed 10 of 10 files at r4, 8 of 11 files at r5, 9 of 9 files at r10, 2 of 2 files at r11, 2 of 2 files at r12.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-:-M9OQcKgDbygWzMqydRk:b-ra0xjj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, @rohany, and @yuzefovich)

___
*[pkg/sql/colexec/hash_utils_tmpl.go, line 62 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9OOBIU1mX3yO6GHdlN:-M9OOBIU1mX3yO6GHdlO:b3c9noi) ([raw file](https://github.com/cockroachdb/cockroach/blob/b37382d67a122a5c689ce03f7e7aed91e5a96138/pkg/sql/colexec/hash_utils_tmpl.go#L62)):*
> ```Go
> 	nKeys int,
> 	sel []int,
> 	hasSel bool,
> ```

I think I agree that having the variable names in the directive makes it easy to notice that these are template variables. Not sure if that's because we're all used to templates. @helenmhe what do you think?
___
*[pkg/sql/colexec/execgen/testdata/inline, line 180 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9OQSnV2Q_PMhSBN2Iv:-M9OQSnV2Q_PMhSBN2Iw:b-x574up) ([raw file](https://github.com/cockroachdb/cockroach/blob/84168bd1eca69e0ed274d2d13879bced2fa01acd/pkg/sql/colexec/execgen/testdata/inline#L180)):*
> ```
> func a() {
> 	{
> 		foo = bar
> ```

This isn't valid go, right?
___
*[pkg/sql/colexec/execgen/testdata/template, line 14 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9OPk_YE8d21Zi1xjsV:-M9OPk_ZBDRc2C35kb_4:b-rx3r11) ([raw file](https://github.com/cockroachdb/cockroach/blob/b37382d67a122a5c689ce03f7e7aed91e5a96138/pkg/sql/colexec/execgen/testdata/template#L14)):*
> ```
> func b(t bool) int {
>   if t {
>     x = 3
> ```

This is slightly confusing. I didn't realize you could reference variables in the outer scope. Wasn't this something you wanted to get rid of?


<!-- Sent from Reviewable.io -->

helenmhe-zz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49939)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, @rohany, and @yuzefovich)

___
*[pkg/sql/colexec/hash_utils_tmpl.go, line 62 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9OOBIU1mX3yO6GHdlN:-M9OTddP6MEX6OxBzMcO:b-5ndw38) ([raw file](https://github.com/cockroachdb/cockroach/blob/b37382d67a122a5c689ce03f7e7aed91e5a96138/pkg/sql/colexec/hash_utils_tmpl.go#L62)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

I think I agree that having the variable names in the directive makes it easy to notice that these are template variables. Not sure if that's because we're all used to templates. @helenmhe what do you think?
</blockquote></details>

Yeah I think I prefer including them in the directive over just the different case


<!-- Sent from Reviewable.io -->

rohany commented on pkg/sql/colexec/execgen/inline.go:
```diff
@@ -54,17 +58,41 @@ func InlineFuncs(inputFileContents string) (string, error) {
 			}
-			decl := getTemplateFunc(templateFuncMap, callExpr)
-			if decl == nil {
+			funcInfo := getTemplateFunc(inlineFuncMap, callExpr)
```
should this still be named getTemplateFunc?

rohany commented on pkg/sql/colexec/execgen/template.go:
```diff
@@ -0,0 +1,351 @@
+		templateArgs[i] = dst.Clone(call.Args[param.fieldOrdinal]).(dst.Expr)
+	}
+	// Remove template vars from callsite.
+	newArgs := make([]dst.Expr, 0, len(call.Args)-len(info.templateParams))
```
That's wierd -- I got panics when I copied the args from the original expression instead of cloning it first, and then removing the other args.

rohany commented on pkg/sql/colexec/execgen/template.go:
```diff
@@ -0,0 +1,351 @@
+			argsPossibilities := generateAllTemplateArgs(info.templateParams)
+
+			funcDecs := info.decl.Decs
+			cursor.InsertBefore(&dst.GenDecl{
```
Could you add a comment here that this is replacing the template func base with a constant? Also, I think this is in conflict with a TODO comment you've left in the inliner.

rohany commented on pkg/sql/colexec/execgen/testdata/template:
```diff
@@ -0,0 +1,142 @@
+inline
+package main
```
I think it would be useful to have the template test run without the inliner steps

rohany reviewed (commented):
I like this approach more, though I feel like templating isn't quite the right word for what's going on here. I can't think of any alternatives though.

yuzefovich reviewed (commented):
> I feel like templating isn't quite the right word for what's going on here. I can't think of any alternatives though.

Some other ideas are: "preprocessing", "precompilation", "pregeneration", "codegeneration". Still I don't think any of these are better than "templating".

> Reviewed 5 of 13 files at r16, 2 of 2 files at r17, 2 of 2 files at r19, 2 of 2 files at r20, 2 of 2 files at r21.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-:-M9Ptegm0tyscjgO_dnO:b7clqvj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto and @jordanlewis)

___
*[pkg/sql/colexec/hashtable_tmpl.go, line 520 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9Kh7v07MOkfRv0vowR:-M9Q3Lq74qFuzLjNbG0o:b-q7k7jj) ([raw file](https://github.com/cockroachdb/cockroach/blob/e919deb166564bece70a23ac8acb957e7c77aaf0/pkg/sql/colexec/hashtable_tmpl.go#L520)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

:thumbsup: 
</blockquote></details>

One idea is to keep the nice variable names but also introduce a convention that all template variables are at the end of the function signature and require some visual separator (for example, `/* runtimeVarsEnd */` before the first template variable when the function signature is on a single line and a separate line with `// runtimeVarsEnd` when it is on multiple lines).
___
*[pkg/sql/colexec/execgen/datadriven_test.go, line 39 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9Q4JRK0lavt8F1O78v:-M9Q4JRK0lavt8F1O78w:b-fl11vs) ([raw file](https://github.com/cockroachdb/cockroach/blob/8245af2fb926833a285d04b6eecd4b8d48424308/pkg/sql/colexec/execgen/datadriven_test.go#L39)):*
> ```Go
> 			}
> 			var sb strings.Builder
> 			_ = decorator.Fprint(&sb, f)
> ```

Is an error ignored here? Maybe we should do `t.Fatal(err)` if non-nil?
___
*[pkg/sql/colexec/execgen/template.go, line 119 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9Kbe5PEd0CpzA-EMS-:-M9Q5NLT-Sf25_PQE2O3:b-kufkx2) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b2ea0fc196f22c9f4788eb2945ca3ada1ef43de/pkg/sql/colexec/execgen/template.go#L119)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

That's how the AST is defined in Go. An If statement always contains a Body (a block stmt) and optionally an Else (a generic stmt). An if /else if / else is actually 2 nested if statements, where the else of the first if contains the second if/else, if that makes sense.

https://astexplorer.net/#/gist/668b53ce1504c5f02ebf56fe778b6717/26076b4bfe01ea86ad0b1b025cdd9a7b0be6fee4
</blockquote></details>

I see, thanks.
___
*[pkg/sql/colexec/execgen/template.go, line 263 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9KdkmV3Y4b8MSzqO8u:-M9Q6rpl2gaYHCO6m8-x:b-wsrk8z) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b2ea0fc196f22c9f4788eb2945ca3ada1ef43de/pkg/sql/colexec/execgen/template.go#L263)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

Good point, I guess I would prefer to get rid of the thing that templateRe does and require a single space. I did that instead.
</blockquote></details>

Just to be clear, by "require a single space" do you mean the writer of the template is responsible for satisfying that requirement?
___
*[pkg/sql/colexec/execgen/template_test.go, line 33 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9Q7KZmEFumNyS0BE-f:-M9Q7KZmEFumNyS0BE-g:bmxdy2p) ([raw file](https://github.com/cockroachdb/cockroach/blob/8245af2fb926833a285d04b6eecd4b8d48424308/pkg/sql/colexec/execgen/template_test.go#L33)):*
> ```Go
> 		argsStr := prettyPrintExprs(args...)
> 		if _, ok := argsMap[argsStr]; ok {
> 			t.Fatalf("Duplicate template args %s", argsStr)
> ```

super nit: we usually use lower case words in the error messages.


<!-- Sent from Reviewable.io -->

jordanlewis reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49939)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, @rohany, and @yuzefovich)

___
*[pkg/sql/colexec/hashtable_tmpl.go, line 520 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9Kh7v07MOkfRv0vowR:-M9QIX8m8noUc78u9bjm:b-iythf0) ([raw file](https://github.com/cockroachdb/cockroach/blob/e919deb166564bece70a23ac8acb957e7c77aaf0/pkg/sql/colexec/hashtable_tmpl.go#L520)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

One idea is to keep the nice variable names but also introduce a convention that all template variables are at the end of the function signature and require some visual separator (for example, `/* runtimeVarsEnd */` before the first template variable when the function signature is on a single line and a separate line with `// runtimeVarsEnd` when it is on multiple lines).
</blockquote></details>

Yeah, maybe. I've backed the changes to individual files out of this PR, so we can debate that in a different one.
___
*[pkg/sql/colexec/execgen/datadriven_test.go, line 39 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9Q4JRK0lavt8F1O78v:-M9QJ9XI9D7VD-N-UYhj:b-f8v3hq) ([raw file](https://github.com/cockroachdb/cockroach/blob/8245af2fb926833a285d04b6eecd4b8d48424308/pkg/sql/colexec/execgen/datadriven_test.go#L39)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Is an error ignored here? Maybe we should do `t.Fatal(err)` if non-nil?
</blockquote></details>

This can never fail, since the writer is a string builder.
___
*[pkg/sql/colexec/execgen/inline.go, line 61 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M996t-dwANG3kVAE2P4-r15-61:-M9PRGD88Lx6lCXcgtxl:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a541256aa986f5effb6fd8afc27381f40eb0a940/pkg/sql/colexec/execgen/inline.go#L61)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

should this still be named getTemplateFunc?
</blockquote></details>

Done.
___
*[pkg/sql/colexec/execgen/template.go, line 263 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9KdkmV3Y4b8MSzqO8u:-M9QJEm_2X270rahrYSS:bb77ufg) ([raw file](https://github.com/cockroachdb/cockroach/blob/6b2ea0fc196f22c9f4788eb2945ca3ada1ef43de/pkg/sql/colexec/execgen/template.go#L263)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Just to be clear, by "require a single space" do you mean the writer of the template is responsible for satisfying that requirement?
</blockquote></details>

Yes.
___
*[pkg/sql/colexec/execgen/template.go, line 242 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9HGeMapB9vKaBifDXu-r15-242:-M9QJFzW0YNvn5LREpbX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a541256aa986f5effb6fd8afc27381f40eb0a940/pkg/sql/colexec/execgen/template.go#L242)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

Could you add a comment here that this is replacing the template func base with a constant? Also, I think this is in conflict with a TODO comment you've left in the inliner.
</blockquote></details>

Done.
___
*[pkg/sql/colexec/execgen/template_test.go, line 33 at r21](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9Q7KZmEFumNyS0BE-f:-M9QJGlEEMaZsziaeD7c:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8245af2fb926833a285d04b6eecd4b8d48424308/pkg/sql/colexec/execgen/template_test.go#L33)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

super nit: we usually use lower case words in the error messages.
</blockquote></details>

Done.
___
*[pkg/sql/colexec/execgen/testdata/inline, line 180 at r11](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9OQSnV2Q_PMhSBN2Iv:-M9PQw7ICvbuXDm4r1eE:b-nkmnb) ([raw file](https://github.com/cockroachdb/cockroach/blob/84168bd1eca69e0ed274d2d13879bced2fa01acd/pkg/sql/colexec/execgen/testdata/inline#L180)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

This isn't valid go, right?
</blockquote></details>

Yeah, just a test of the inliner/templater which doesn't check Go code validity.
___
*[pkg/sql/colexec/execgen/testdata/template, line 14 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9OPk_YE8d21Zi1xjsV:-M9PQmOS5BzV-i5Zhex6:bt6gipm) ([raw file](https://github.com/cockroachdb/cockroach/blob/b37382d67a122a5c689ce03f7e7aed91e5a96138/pkg/sql/colexec/execgen/testdata/template#L14)):*
<details><summary><i>Previously, asubiotto (Alfonso Subiotto Marqués) wrote…</i></summary><blockquote>

This is slightly confusing. I didn't realize you could reference variables in the outer scope. Wasn't this something you wanted to get rid of?
</blockquote></details>

Can't do this easily. These tests don't actually get compiled, it's just to test the templating/inlining - wouldn't matter if this was x, or someVarThatDoesntExist.
___
*[pkg/sql/colexec/execgen/testdata/template, line 2 at r15](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-M9HGeMapB9vKaBifDXw-r15-2:-M9PQlZaElyLGcBVFWBH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a541256aa986f5effb6fd8afc27381f40eb0a940/pkg/sql/colexec/execgen/testdata/template#L2)):*
<details><summary><i>Previously, rohany (Rohan Yadav) wrote…</i></summary><blockquote>

I think it would be useful to have the template test run without the inliner steps
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

jordanlewis commented:
FYI, I moved all the individual template file changes to #50030.

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 20 of 20 files at r22, 9 of 9 files at r23, 2 of 2 files at r24.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/49939#-:-M9QaOeF4jcykakGbZog:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @asubiotto, @jordanlewis, and @rohany)



<!-- Sent from Reviewable.io -->

jordanlewis commented:
TFTRs!

jordanlewis commented:
bors r+
